src/test/java/io/cryostat/rules/RuleProcessorTest.java

```diff
@@ -377,4 +382,46 @@ void testTaskCancellationOnFailure() throws Exception {
 
         Mockito.verify(vertx).cancelTimer(MockVertx.TIMER_ID);
     }
+
+    @Test
+    void testEventCallOnCredentialsChange() throws Exception {
+
+        Credentials credentials = new Credentials("foouser", "barpassword");
+        String jmxUrl = "service:jmx:rmi://localhost:9091/jndi/rmi://fooHost:9091/jmxrmi";
+        ServiceRef serviceRef = new ServiceRef(new URI(jmxUrl), "com.example.App");
+        String matchExpression = "target.alias == 'com.example.App'";
+
+        Event<CredentialsEvent, String> event = Mockito.mock(Event.class);
+        Mockito.when(event.getEventType()).thenReturn(CredentialsEvent.ADDED);
+        Mockito.when(credentialsManager.resolveMatchingTargets(event.getPayload()))
+                .thenReturn(Set.of(serviceRef));
+
+        Rule rule =
+                new Rule.Builder()
+                        .name("Test Rule")
+                        .description("Automated unit test rule")
+                        .matchExpression(matchExpression)
+                        .eventSpecifier("archive")
+                        .build();
+
+        Mockito.when(registry.getRules(serviceRef)).thenReturn(Set.of(rule));
+
+        EventListener<CredentialsManager.CredentialsEvent, String> listener =
+                processor.credentialsListener();
+        CredentialsManager credentialsManager = Mockito.mock(CredentialsManager.class);
+
+        Mockito.doAnswer(
+                        new Answer<Void>() {
+
+                            @Override
+                            public Void answer(InvocationOnMock invocation) throws Throwable {
+                                listener.onEvent(event);
+                                return null;
+                            }
+                        })
+                .when(credentialsManager)
+                .addCredentials(matchExpression, credentials);
+
+        credentialsManager.addCredentials(matchExpression, credentials);
```

Comment:
Could you add some `Mockito.verify()` to check behaviours, or `MatcherAssert.assertThat()` to check that values passed/returned are all as expected?

In another similar test, everything from this line to the end of the method is basically verifying that the expected behaviour occurred:

https://github.com/cryostatio/cryostat/blob/a8c706207f48c991bc9c7055c9131068347a0751/src/test/java/io/cryostat/rules/RuleProcessorTest.java#L201

A check like this:

```java
        Mockito.verify(recordingTargetHelper)
                .startRecording(
                        restartCaptor.capture(),
                        connectionDescriptorCaptor.capture(),
                        recordingOptionsCaptor.capture(),
                        templateNameCaptor.capture(),
                        templateTypeCaptor.capture(),
                        metadataCaptor.capture());

        Assertions.assertTrue(restartCaptor.getValue());
```

looks like it would be good to have here, too.

Reply:
- Added assert to check if `addCredentials` return is greater than -1 as it returns the credential id (>0). 
- Added a verify to check if the startRecording is called.